parser-typechecker/src/Unison/Codebase/SqliteCodebase.hs

```diff
@@ -1077,40 +1109,38 @@ pushGitRootBranch srcConn branch repo = runExceptT @C.GitError do
 
   -- set up the cache dir
   remotePath <- time "Git fetch" $ withExceptT C.GitProtocolError $ pullBranch (writeToRead repo)
-  destConn <- openOrCreateCodebaseConnection "push.dest" remotePath
-
-  flip runReaderT destConn $ Q.savepoint "push"
-  lift . flip State.execStateT emptySyncProgressState $
-    syncInternal syncProgress srcConn destConn (Branch.transform lift branch)
-  flip runReaderT destConn do
-    let newRootHash = Branch.headHash branch
-    -- the call to runDB "handles" the possible DB error by bombing
-    (fmap . fmap) Cv.branchHash2to1 (runDB destConn Ops.loadMaybeRootCausalHash) >>= \case
-      Nothing -> do
-        setRepoRoot newRootHash
-        Q.release "push"
-      Just oldRootHash -> do
-        before oldRootHash newRootHash >>= \case
-          Nothing ->
-            error $
-              "I couldn't find the hash " ++ show newRootHash
-                ++ " that I just synced to the cached copy of "
-                ++ repoString
-                ++ " in "
-                ++ show remotePath
-                ++ "."
-          Just False -> do
-            Q.rollbackRelease "push"
-            throwError . C.GitProtocolError $ GitError.PushDestinationHasNewStuff repo
-
-          Just True -> do
-            setRepoRoot newRootHash
-            Q.release "push"
-
-    Q.setJournalMode JournalMode.DELETE
-
+  (closeDestConn, destConn) <- openOrCreateCodebaseConnection "push.dest" remotePath
+  mapExceptT (`finally` liftIO closeDestConn) $ do
```

Comment:
Here we interleave IO with ExceptT and it gets a bit annoying, but in the existing code we don't correctly close the connection on either `Left`s or on IO exceptions, this should handle both.

We can't use `withConnection` here because ExceptT isn't a valid `MonadUnliftIO`

Reply:
Nice, if/when I can find time to continue the transaction work we can eliminate the `ExceptT` layer induced by `Queries`/`Operations`.